test/shmemx/shmem_malloc_with_hints.c

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Copyright (c) 2022 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <shmem.h>
+#include <shmemx.h>
+
+#define N 128
+#define SHMEM_MALLOC_INVALID_HINT ~(SHMEM_MALLOC_ATOMICS_REMOTE)
+
+int main(int argc, char **argv) {
+    int npes, mype;
+    int errors = 0;
+
+    shmem_init();
+
+    npes = shmem_n_pes();
+    mype = shmem_my_pe();
+
+    if (npes < 2) {
+        printf("This test requires at least 2 PEs\n");
+        shmem_global_exit(0);
+    }
+
+    if (npes > 2 && mype == 0) {
+        printf("This test is written for 2 PEs, only PE0 and PE1 will perform the data transfer\n");
+    }
+
+    int *src[N];
+
+    /* Allocate an array of N buffers on the symmeytric heap */
+    for (int i = 0; i < N; i++)
+        src[i] = (int *)shmem_malloc_with_hints(N * sizeof(int), SHMEMX_MALLOC_NO_BARRIER);
+    int *dst = (int *)shmem_malloc(N * sizeof(int));
+
+    for (int i = 0; i < N; i++) {
+        for (int j = 0; j < N; j++) {
+            src[i][j] = -1;
+        }
+    }
+
+    /* src is initialized to become a diagonal matrix */
+    for (int i = 0; i < N; i++) {
+        src[i][i] = i;
+        dst[i] = -1;
+    }
+
+    shmem_sync_all(); /* sync sender and receiver */
+
+    if (mype == 0) {
+        for (int i = 0; i < N; i++) {
+            /* put elements from src's diagonal into dst on PE 1 */
+            shmem_int_put(&dst[i], &src[i][i], 1, 1);
```

Comment:
Perhaps:
```suggestion
            shmem_int_put(&dst[i], &src[i][i], 1, i % npes);
```
might as well exercise all PEs, right?

Reply:
I want to make sure that I understand the change that you're suggesting. Are you saying that on each iteration I should put the `src` value from PE0 into `dst` on either `PE0`, `PE1`, ..., `PE<N>` and then verify that the correct values across each PE's `dst` was  `put` into their appropriate index by the end?

Would it also satisfy if PE0 did a `get` from each of the other PE's (`0` - `N`'s) `dst`? And then just verfiy that `dst` is correct within `PE0` at the end? 
```suggestion
            shmem_int_get(&dst[i], &src[i][i], 1, i % npes);
```

Reply:
Yes, exactly.  And sure, I think the Get version makes a little more sense.  One funny thing there is that `dst` would not need to be symmetric.  So maybe you could use a `shmem_barrier_all()` and `malloc` instead of `shmem_malloc` for `dst`.  (also I clarified the last sentence in the SHMEMX wiki to account for this option)

Reply:
Awesome, I just updated the code. Thanks for the suggestion. and the clarification in the SHMEMX wiki was very helpful, thank you!